linkerd/io/src/scoped.rs

```diff
@@ -0,0 +1,133 @@
+use crate as io;
+use pin_project::pin_project;
+use std::{pin::Pin, task::Context};
+
+/// An I/O stream where errors are annotated a scope.
+#[pin_project]
+#[derive(Debug)]
+pub struct ScopedIo<I> {
+    scope: Scope,
+
+    #[pin]
+    io: I,
+}
+
+#[derive(Copy, Clone, Debug)]
+enum Scope {
+    Client,
+    Server,
+}
+
+// === impl Scope ===
+
+impl Scope {
+    fn err(&self, err: io::Error) -> io::Error {
+        io::Error::new(err.kind(), format!("{}: {}", self, err.to_string()))
+    }
+}
+
+impl std::fmt::Display for Scope {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Client => write!(f, "client"),
+            Self::Server => write!(f, "server"),
+        }
+    }
+}
+
+// === impl ScopedIo ===
+
+impl<I> ScopedIo<I> {
+    pub fn client(io: I) -> Self {
+        Self {
+            scope: Scope::Client,
+            io,
+        }
+    }
+
+    pub fn server(io: I) -> Self {
+        Self {
+            scope: Scope::Server,
+            io,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl<I: io::Peek + Send + Sync> io::Peek for ScopedIo<I> {
+    async fn peek(&self, buf: &mut [u8]) -> io::Result<usize> {
+        self.io.peek(buf).await.map_err(|e| self.scope.err(e))
```

Comment:
nit/tioli: we _could_ change `self.scope.err()` to return a closure, and then write 
```rust
.map_err(self.scope.err())
```
here and everywhere else. maybe slightly more concise. but, that might also just be confusing, so idk.